File: packages/dart/bowline/test/transport_test.dart
import 'dart:async';
import 'dart:convert';

import 'package:bowline/bowline.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

Transport transport(MockClient client,
        {Map<String, String> Function()? headers}) =>
    Transport(
        baseUrl: Uri.parse('http://api.test/api/'),
        client: client,
        headers: headers);

void main() {
  test('GET queries carry the input as a query parameter', () async {
    late http.Request seen;
    final t = transport(MockClient((request) async {
      seen = request;
      return http.Response('{"id":3,"name":"Ada"}', 200);
    }), headers: () => {'authorization': 'Bearer static'});
    final out = await t.call(
      'users.get',
      Method.get,
      {'id': 3},
      (json) => asObject(json)['name'],
      options: const CallOptions(headers: {'x-tenant': 'acme'}),
    );
    expect(out, 'Ada');
    expect(seen.method, 'GET');
    expect(seen.url.path, '/api/users.get');
    expect(seen.url.queryParameters['input'], '{"id":3}');
    expect(seen.headers['authorization'], 'Bearer static');
    expect(seen.headers['x-tenant'], 'acme');
    expect(seen.headers['accept'], 'application/json');
  });

  test('POST mutations send a JSON body with the content type', () async {
    late http.Request seen;
    final t = transport(MockClient((request) async {
      seen = request;
      return http.Response('{"id":9}', 200);
    }));
    await t.call('users.create', Method.post, {'name': 'Grace'}, asRaw);
    expect(seen.method, 'POST');
    expect(seen.body, '{"name":"Grace"}');
    expect(seen.headers['content-type'], startsWith('application/json'));
  });

  test('error envelopes become BowlineException', () async {
    final t = transport(MockClient((request) async => http.Response(
          '{"error":{"code":"NOT_FOUND","message":"user 9 not found","issues":[{"path":["id"],"rule":"exists","message":"missing"}]}}',
          404,
        )));
    try {
      await t.call('users.get', Method.get, {'id': 9}, asRaw);
      fail('expected an exception');
    } on BowlineException catch (e) {
      expect(e.code, Code.notFound);
      expect(e.status, 404);
      expect(e.message, 'user 9 not found');
      expect(e.issues, [
        const Issue(['id'], 'exists', 'missing')
      ]);
    }
  });

  test('non-envelope failures map to UNKNOWN with the status', () async {
    final t = transport(
        MockClient((request) async => http.Response('gateway down', 502)));
    await expectLater(
      t.call('users.get', Method.get, {}, asRaw),
      throwsA(isA<BowlineException>()
          .having((e) => e.code, 'code', Code.unknown)
          .having((e) => e.status, 'status', 502)),
    );
  });

  test('network failures map to UNAVAILABLE with status 0', () async {
    final t = transport(
        MockClient((request) async => throw http.ClientException('refused')));
    await expectLater(
      t.call('users.get', Method.get, {}, asRaw),
      throwsA(isA<BowlineException>()
          .having((e) => e.code, 'code', Code.unavailable)
          .having((e) => e.status, 'status', 0)),
    );
  });

  test('timeouts map to DEADLINE_EXCEEDED', () async {
    final t = transport(MockClient((request) async {
      await Future<void>.delayed(const Duration(milliseconds: 200));
      return http.Response('{}', 200);
    }));
    await expectLater(
      t.call('users.get', Method.get, {}, asRaw,
          options: const CallOptions(timeout: Duration(milliseconds: 20))),
      throwsA(isA<BowlineException>()
          .having((e) => e.code, 'code', Code.deadlineExceeded)),
    );
  });

  test('subscriptions yield message events and stop at done', () async {
    final t = transport(MockClient((request) async {
      expect(request.headers['accept'], 'text/event-stream');
      const body = ': open\n\n'
          'event: message\ndata: {"n":1}\n\n'
          ': ping\n\n'
          'event: message\ndata: {"n":2}\n\n'
          'event: done\ndata: {}\n\n'
          'event: message\ndata: {"n":3}\n\n';
      return http.Response(body, 200,
          headers: {'content-type': 'text/event-stream'});
    }));
    final seen = await t
        .subscribe('ticks', {'count': 2}, (json) => asObject(json)['n'])
        .toList();
    expect(seen, [1, 2]);
  });

  test('subscription error events throw the envelope', () async {
    final t = transport(MockClient((request) async => http.Response(
          'event: message\ndata: {"n":1}\n\nevent: error\ndata: {"error":{"code":"PERMISSION_DENIED","message":"no"}}\n\n',
          200,
        )));
    final stream = t.subscribe('ticks', {}, (json) => asObject(json)['n']);
    await expectLater(
      stream.toList(),
      throwsA(isA<BowlineException>()
          .having((e) => e.code, 'code', Code.permissionDenied)),
    );
  });

  test('uploads send the input part followed by the file part', () async {
    late http.MultipartRequest seen;
    final t = transport(MockClient.streaming((request, bodyStream) async {
      seen = request as http.MultipartRequest;
      final body = await http.ByteStream(bodyStream).bytesToString();
      expect(
          body.indexOf('name="input"'), lessThan(body.indexOf('name="file"')));
      expect(body, contains('{"label":"receipt"}'));
      expect(body, contains('hello'));
      return http.StreamedResponse(
          Stream.value(utf8.encode('{"size":5}')), 200);
    }));
    final out = await t.upload(
      'upload',
      {'label': 'receipt'},
      Stream.value(utf8.encode('hello')),
      'receipt.bin',
      (json) => asObject(json)['size'],
    );
    expect(out, 5);
    expect(seen.files.map((f) => f.field), ['input', 'file']);
    expect(seen.files.last.filename, 'receipt.bin');
  });
}

File: packages/elixir/bowline_client/test/transport_test.exs
defmodule BowlineClient.TransportTest do
  use ExUnit.Case, async: true

  alias BowlineClient.Error
  alias BowlineClient.TestServer
  alias BowlineClient.Transport

  defp serve(handler) do
    server = {url, _, _} = TestServer.start(handler)
    on_exit(fn -> TestServer.stop(server) end)
    url
  end

  test "GET queries carry the input as a query parameter and static plus per-call headers" do
    url = serve(fn _ -> {200, ~s({"id":3,"name":"Ada"})} end)

    transport =
      Transport.new(base_url: url <> "/api/", headers: fn -> %{"authorization" => "Bearer t"} end)

    assert {:ok, %{"id" => 3}} =
             Transport.call(transport, "users.get", :get, %{"id" => 3}, & &1,
               headers: %{"x-tenant" => "acme"}
             )

    assert_receive {:request, request}
    assert request.method == "GET"
    assert request.path == "/api/users.get"
    assert request.query == %{"input" => ~s({"id":3})}
    assert request.headers["authorization"] == "Bearer t"
    assert request.headers["x-tenant"] == "acme"
    assert request.headers["accept"] == "application/json"
  end

  test "POST mutations send a JSON body" do
    url = serve(fn _ -> {200, ~s({"ok":true})} end)
    transport = Transport.new(base_url: url)

    assert {:ok, %{"ok" => true}} =
             Transport.call(transport, "users.create", :post, %{"name" => "x"}, & &1)

    assert_receive {:request, request}
    assert request.method == "POST"
    assert request.headers["content-type"] == "application/json"
    assert JSON.decode!(request.body) == %{"name" => "x"}
  end

  test "an error envelope becomes a BowlineClient.Error with code, status, details, and issues" do
    body =
      ~s({"error":{"code":"INVALID_ARGUMENT","message":"invalid input","issues":[{"path":["lines","0","quantity"],"rule":"min","message":"must be at least 1"}]}})

    url = serve(fn _ -> {400, body} end)
    transport = Transport.new(base_url: url)
    assert {:error, %Error{} = error} = Transport.call(transport, "x", :post, %{}, & &1)
    assert error.code == :invalid_argument
    assert error.status == 400
    assert error.message == "invalid input"
    assert [%BowlineClient.Issue{path: ["lines", "0", "quantity"], rule: "min"}] = error.issues
    assert Exception.message(error) =~ "lines.0.quantity: must be at least 1"
  end

  test "a non-envelope failure maps to unknown with the status" do
    url = serve(fn _ -> {502, [{"content-type", "text/html"}], "<h1>bad gateway</h1>"} end)
    transport = Transport.new(base_url: url)

    assert {:error, %Error{code: :unknown, status: 502}} =
             Transport.call(transport, "x", :get, %{}, & &1)
  end

  test "a network failure maps to unavailable with status 0" do
    {url, _, listen} = TestServer.start(fn _ -> {200, "{}"} end)
    :gen_tcp.close(listen)
    transport = Transport.new(base_url: url)

    assert {:error, %Error{code: :unavailable, status: 0}} =
             Transport.call(transport, "x", :get, %{}, & &1)
  end

  test "a timeout maps to deadline_exceeded" do
    url = serve(fn _ -> {:delay, 300, {200, "{}"}} end)
    transport = Transport.new(base_url: url)

    assert {:error, %Error{code: :deadline_exceeded}} =
             Transport.call(transport, "x", :get, %{}, & &1, timeout: 50)
  end

  test "decoder failures surface as internal errors instead of raising" do
    url = serve(fn _ -> {200, ~s({"id":"three"})} end)
    transport = Transport.new(base_url: url)

    assert {:error, %Error{code: :internal, message: message}} =
             Transport.call(transport, "x", :get, %{}, &BowlineClient.Read.int(&1, "id"))

    assert message =~ "id: expected integer"
  end

  test "subscriptions stream decoded messages until done" do
    chunks = [
      ": open\n\n",
      "event: message\ndata: {\"n\":1}\n\n",
      ": ping\n\nevent: message\ndata: {\"n\":2}\n\nevent: mess",
      "age\ndata: {\"n\":3}\n\nevent: done\ndata: {}\n\n"
    ]

    url = serve(fn _ -> {:stream, chunks} end)
    transport = Transport.new(base_url: url)
    assert {:ok, stream} = Transport.subscribe(transport, "ticks", %{"count" => 3}, & &1["n"])
    assert Enum.to_list(stream) == [1, 2, 3]
    assert_receive {:request, request}
    assert request.headers["accept"] == "text/event-stream"
    assert request.query == %{"input" => ~s({"count":3})}
  end

  test "a subscription error event raises the envelope" do
    chunks = ["event: error\ndata: {\"error\":{\"code\":\"NOT_FOUND\",\"message\":\"gone\"}}\n\n"]
    url = serve(fn _ -> {:stream, chunks} end)
    transport = Transport.new(base_url: url)
    assert {:ok, stream} = Transport.subscribe(transport, "ticks", %{}, & &1)
    assert_raise Error, fn -> Enum.to_list(stream) end
  end

  test "uploads send the input part before the file part" do
    url = serve(fn _ -> {200, ~s({"size":5})} end)
    transport = Transport.new(base_url: url)

    assert {:ok, %{"size" => 5}} =
             Transport.upload(
               transport,
               "files.put",
               %{"label" => "x"},
               ["hel", "lo"],
               "a.bin",
               & &1
             )

    assert_receive {:request, request}
    assert request.headers["content-type"] =~ "multipart/form-data"
    [_, input_part, file_part | _] = String.split(request.body, "--")
    assert input_part =~ ~s(name="input")
    assert input_part =~ ~s({"label":"x"})
    assert file_part =~ ~s(name="file")
    assert file_part =~ ~s(filename="a.bin")
    assert file_part =~ "hello"
  end
end
